### ISSUE TITLE
Using ModelState and Other ControllerBase Properties Outside of a Controller

### ISSUE THREAD

**Issue by pdevito3** [jbogard/MediatR#514]
I'm working on moving my API logic in my PATCH endpoint to a Mediatr Command. When applying my patch document, I usually check the model state like below. Normally, I'm doing this from a controller so there is no issue, but when moving this into a RequestHandler, I no longer have access to the model state property since I'm outside of the controller. How would you recommend going about this?

The model state logic I'd like to use outside of the controller:
```csharp
updatePartialValueToReplaceCommand.PatchDoc.ApplyTo(valueToReplaceToPatch, ModelState); // apply patchdoc updates to the updatable valueToReplace

if (!TryValidateModel(valueToReplaceToPatch))
{
    return ValidationProblem(ModelState);
}
```

The rest of the code for context:

**Patch Endpoint**
```csharp

        [HttpPatch("{valueToReplaceId}")]
        public IActionResult PartiallyUpdateValueToReplace(int valueToReplaceId, JsonPatchDocument<ValueToReplaceForUpdateDto> patchDoc)
        {
            var query = new UpdatePartialValueToReplaceCommand(valueToReplaceId, patchDoc);
            var result = _mediator.Send(query);

            switch (result.Result.ToUpper())
            {
                case "NOTFOUND":
                    return NotFound();
                case "NOCONTENT":
                    return NoContent();
                default:
                    return BadRequest();
            }
        }
```

**UpdatePartialValueToReplaceCommand** 
```csharp
public class UpdatePartialValueToReplaceCommand : IRequest<string>
    {
        public int ValueToReplaceId { get; set; }
        public JsonPatchDocument<ValueToReplaceForUpdateDto> PatchDoc { get; set; }

        public UpdatePartialValueToReplaceCommand(int valueToReplaceId, JsonPatchDocument<ValueToReplaceForUpdateDto> patchDoc)
        {
            ValueToReplaceId = valueToReplaceId;
            PatchDoc = patchDoc;
        }
    }
```

**(BROKEN) UpdatePartialValueToReplaceHandler** 
```csharp
    public class UpdatePartialValueToReplaceHandler : IRequestHandler<UpdatePartialValueToReplaceCommand, string>
    {
        private readonly IValueToReplaceRepository _valueToReplaceRepository;
        private readonly IMapper _mapper;

        public UpdatePartialValueToReplaceHandler(IValueToReplaceRepository valueToReplaceRepository
            , IMapper mapper)
        {
            _valueToReplaceRepository = valueToReplaceRepository ??
                throw new ArgumentNullException(nameof(valueToReplaceRepository));
            _mapper = mapper ??
                throw new ArgumentNullException(nameof(mapper));
        }

        public async Task<string> Handle(UpdatePartialValueToReplaceCommand updatePartialValueToReplaceCommand, CancellationToken cancellationToken)
        {
            if (updatePartialValueToReplaceCommand.PatchDoc == null)
            {
                return "BadRequest";
            }

            var existingValueToReplace = _valueToReplaceRepository.GetValueToReplace(updatePartialValueToReplaceCommand.ValueToReplaceId);

            if (existingValueToReplace == null)
            {
                return "NotFound";
            }

            var valueToReplaceToPatch = _mapper.Map<ValueToReplaceForUpdateDto>(existingValueToReplace); // map the valueToReplace we got from the database to an updatable valueToReplace model
            updatePartialValueToReplaceCommand.PatchDoc.ApplyTo(valueToReplaceToPatch, ModelState); // apply patchdoc updates to the updatable valueToReplace -- THIS DOESN'T WORK IN A MEDIATR COMMAND BECAUSE I DON'T HAVE CONTROLLERBASE CONTEXT

            if (!TryValidateModel(valueToReplaceToPatch))
            {
                return ValidationProblem(ModelState);
            }

            _mapper.Map(valueToReplaceToPatch, existingValueToReplace); // apply updates from the updatable valueToReplace to the db entity so we can apply the updates to the database
            _valueToReplaceRepository.UpdateValueToReplace(existingValueToReplace); // apply business updates to data if needed

            _valueToReplaceRepository.Save(); // save changes in the database

            return "NoContent";
        }
    }
```

**Comment by d1820** [comment]
you could pass a func to the UpdatePartialValueToReplaceCommand to execute the validate in the context of the controller. you also prob dont need to pass back out strings for types of Results to return, the exception middleware can handle that instead

**Comment by pdevito3** [comment]
Man, it's amazing what fresh eyes after a few days will do! Also, your idea got me thinking a bit more and I was finally able to figure out the config to get everything where it needs to be. And got rid of those annoying magic strings I hated as well! Here's the code if you're interested. Appreciate the time!

```csharp
[HttpPatch("{valueToReplaceId}")]
        public IActionResult PartiallyUpdateValueToReplace(int valueToReplaceId, JsonPatchDocument<ValueToReplaceForUpdateDto> patchDoc)
        {
            var query = new UpdatePartialValueToReplaceCommand(valueToReplaceId, patchDoc, this);
            var result = _mediator.Send(query);

            return result.Result;
        }
```

```csharp
    public class UpdatePartialValueToReplaceCommand : IRequest<IActionResult>
    {
        public int ValueToReplaceId { get; set; }
        public JsonPatchDocument<ValueToReplaceForUpdateDto> PatchDoc { get; set; }
        public Controller Controller { get; set; }

        public UpdatePartialValueToReplaceCommand(int valueToReplaceId, JsonPatchDocument<ValueToReplaceForUpdateDto> patchDoc,
            Controller controller)
        {
            ValueToReplaceId = valueToReplaceId;
            PatchDoc = patchDoc;
            Controller = controller;
        }
    }
```

```csharp
public class UpdatePartialValueToReplaceHandler : IRequestHandler<UpdatePartialValueToReplaceCommand, IActionResult>
    {
        private readonly IValueToReplaceRepository _valueToReplaceRepository;
        private readonly IMapper _mapper;

        public UpdatePartialValueToReplaceHandler(IValueToReplaceRepository valueToReplaceRepository
            , IMapper mapper)
        {
            _valueToReplaceRepository = valueToReplaceRepository ??
                throw new ArgumentNullException(nameof(valueToReplaceRepository));
            _mapper = mapper ??
                throw new ArgumentNullException(nameof(mapper));
        }

        public async Task<IActionResult> Handle(UpdatePartialValueToReplaceCommand updatePartialValueToReplaceCommand, CancellationToken cancellationToken)
        {
            if (updatePartialValueToReplaceCommand.PatchDoc == null)
            {
                return updatePartialValueToReplaceCommand.Controller.BadRequest();
            }

            var existingValueToReplace = _valueToReplaceRepository.GetValueToReplace(updatePartialValueToReplaceCommand.ValueToReplaceId);

            if (existingValueToReplace == null)
            {
                return updatePartialValueToReplaceCommand.Controller.NotFound();
            }

            var valueToReplaceToPatch = _mapper.Map<ValueToReplaceForUpdateDto>(existingValueToReplace); // map the valueToReplace we got from the database to an updatable valueToReplace model
            updatePartialValueToReplaceCommand.PatchDoc.ApplyTo(valueToReplaceToPatch, updatePartialValueToReplaceCommand.Controller.ModelState); // apply patchdoc updates to the updatable valueToReplace

            if (!updatePartialValueToReplaceCommand.Controller.TryValidateModel(valueToReplaceToPatch))
            {
                return updatePartialValueToReplaceCommand.Controller.ValidationProblem(updatePartialValueToReplaceCommand.Controller.ModelState);
            }

            _mapper.Map(valueToReplaceToPatch, existingValueToReplace); // apply updates from the updatable valueToReplace to the db entity so we can apply the updates to the database
            _valueToReplaceRepository.UpdateValueToReplace(existingValueToReplace); // apply business updates to data if needed

            _valueToReplaceRepository.Save(); // save changes in the database

            return updatePartialValueToReplaceCommand.Controller.NoContent();
        }
    }
```

**Comment by d1820** [comment]
Only issues with that could be testing may be a pain, to setup a mock controller with all the stuff. vs just having to test if a certain type of exception if thrown, or if a mock validateFunc was called.. But otherwise glad you could move forward

**Comment by pdevito3** [comment]
Yeah, that’s what I’m planning look at in the next couple days. Hoping that I can use Moq and call it a day, but we’ll see how it goes. Will post an update here

**Comment by lilasquared** [comment]
@pdevito3 Just curious but what value to you get in moving the validation to the handler when you're just using the controller model state validation anyway?

**Comment by pdevito3** [comment]
> @pdevito3 Just curious but what value to you get in moving the validation to the handler when you're just using the controller model state validation anyway?

If I have it all in the handler, it will (hopefully) simply my testing and keep my logic separated out. That’s the goal at least!

**Comment by pdevito3** [comment]
Okay, so as mentioned, the implementation works fine in practice, but the testing unfortunately didn't go smoothly here. 

The issue is that when making a `new controller()` or a Mock of a controller, the necessary properties (e.g. modelstate) do not get added properly and don't allow for normal operations. From what I can tell, there might be a convoluted way to mock this up using the `.Setup()` method, but I can't get it nailed down. 

@d1820, I tried briefly to pass in a function to execute the controller as you suggested, but also had no luck here. Not sure I fully understood what you were getting at though, so if you could provide a brief example I could give it another go!

Bottom line, bringing the controller in for validation works, but is very probably a code smell of some kind. Would be great to pass the in a function or validate the patch against the domain like any 'complete' object, but that's not in the cards at the moment.

**Comment by d1820** [comment]
off the cuff would look something like this.


in controller
```
var response = await mediator.Send(UpdatePartialValueToReplaceCommand);
if(response == CustomValidationCode)
{
 return ValidationProblem(ModelState);
}
```
```


public UpdatePartialValueToReplaceCommand(int valueToReplaceId, JsonPatchDocument<ValueToReplaceForUpdateDto> patchDoc,
            Func<bool> validator)
        {
            ValueToReplaceId = valueToReplaceId;
            PatchDoc = patchDoc;
            Validator = validator;
        }


public async Task<IActionResult> Handle(UpdatePartialValueToReplaceCommand updatePartialValueToReplaceCommand, CancellationToken cancellationToken)
        {
            if (updatePartialValueToReplaceCommand.PatchDoc == null)
            {
                    throw new CustomValidationException("message") // this would handled in exception handler, see documents for example
 OR
return HttpStautsCode.BadRequest;
            }

            var existingValueToReplace = _valueToReplaceRepository.GetValueToReplace(updatePartialValueToReplaceCommand.ValueToReplaceId);

            if (existingValueToReplace == null)
            {
                throw new CustomValidationException("message") // this would handled in exception handler, see documents for example
 OR
return HttpStautsCode.BadRequest;
            }

            var valueToReplaceToPatch = _mapper.Map<ValueToReplaceForUpdateDto>(existingValueToReplace); // map the valueToReplace we got from the database to an updatable valueToReplace model
            updatePartialValueToReplaceCommand.PatchDoc.ApplyTo(valueToReplaceToPatch, updatePartialValueToReplaceCommand.Controller.ModelState); // apply patchdoc updates to the updatable valueToReplace

            if (!updatePartialValueToReplaceCommand.Controller.TryValidateModel(valueToReplaceToPatch))
            {
return CustomValidationCode //int
             
            }

            _mapper.Map(valueToReplaceToPatch, existingValueToReplace); // apply updates from the updatable valueToReplace to the db entity so we can apply the updates to the database
            _valueToReplaceRepository.UpdateValueToReplace(existingValueToReplace); // apply business updates to data if needed

            _valueToReplaceRepository.Save(); // save changes in the database

            return HttpStatusCode.OK;
        }
```

I have a whole entire Controller  builder somewhere that was create to moq and test all aspects of a .net controller. If i get some time i can get a gist together of it.

**Comment by pdevito3** [comment]
This is great! I'll take a crack at implementing in the next couple days. 

Appreciate you taking the time! Will keep an eye out for the gist.

**Comment by d1820** [comment]
Files needed to be able to Moq Controllers, and HttpContext. A lot of the .Net Core stuff already exposes way to mock some of it, but at the time they did not provide all the layers needed, in an easy and usable way.

https://gist.github.com/d1820/b4fffe65f26f86a8b1057949310b0d14

**Comment by pdevito3** [comment]
@d1820, that gist is fantastic. Very much appreciated. 

I started looking at it last night and got the basics working, but I'm going to need to tweak it to get the `tryvalidatemodel()` method to work as well as the `Link.Url` part for my get collection query, but this is a great base to work from. Thanks again for taking the time.
